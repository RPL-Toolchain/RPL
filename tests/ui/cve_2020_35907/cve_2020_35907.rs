use core::task::{RawWaker, RawWakerVTable, Waker};
use std::cell::UnsafeCell;

fn noop_waker() -> Waker {
    unsafe fn clone(_data: *const ()) -> RawWaker {
        RawWaker::new(std::ptr::null(), &NOOP_WAKER_VTABLE)
    }

    unsafe fn wake(_data: *const ()) {}

    unsafe fn wake_by_ref(_data: *const ()) {}

    unsafe fn drop(_data: *const ()) {}

    static NOOP_WAKER_VTABLE: RawWakerVTable = RawWakerVTable::new(clone, wake, wake_by_ref, drop);

    unsafe { Waker::from_raw(RawWaker::new(std::ptr::null(), &NOOP_WAKER_VTABLE)) }
}

// #[rpl::dump_mir(dump_cfg, dump_ddg)]
pub fn noop_waker_ref() -> &'static Waker {
    //~^ ERROR: it is unsound to expose a `&'static std::task::Waker` from a thread-local where `std::task::Waker` is `Sync`
    thread_local! {
        static NOOP_WAKER_INSTANCE: UnsafeCell<Waker> =
            UnsafeCell::new(noop_waker());
    }
    NOOP_WAKER_INSTANCE.with(|l| unsafe { &*l.get() })
}
