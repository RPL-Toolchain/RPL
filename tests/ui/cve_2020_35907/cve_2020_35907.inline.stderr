error: it is unsound to expose a `&'static std::task::Waker` from a thread-local where `std::task::Waker` is `Sync`
  --> tests/ui/cve_2020_35907/cve_2020_35907.rs:25:28
   |
LL | pub fn noop_waker_ref() -> &'static Waker {
   |                            ^^^^^^^^^^^^^^ function returning `&'static std::task::Waker` here
...
LL |     NOOP_WAKER_INSTANCE.with(|l| unsafe { &*l.get() })
   |     ------------------- ------------------------------ `&'static std::task::Waker` returned here
   |     |
   |     thread local used here
   |
   = help: `std::task::Waker` is `Sync` so that it can shared among threads
   = help: the thread local is destroyed after the thread has been destroyed, and the exposed `&'static std::task::Waker` may outlive the thread it is exposed to
   = note: `#[deny(rpl::thread_local_static_ref)]` on by default

error: aborting due to 1 previous error

